sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -516,29 +542,60 @@ pi_result piextGetDeviceFunctionPointer(pi_device device, pi_program program,
-      reinterpret_cast<FuncT>(clGetExtensionFunctionAddressForPlatform(
-          cast<cl_platform_id>(platform), "clGetDeviceFunctionPointerINTEL"));
-  // TODO: once we have check that device supports corresponding extension,
-  // we can insert an assertion that func_ptr is not nullptr. For now, let's
-  // just return an error if failed to query such function
-  // assert(
-  //     func_ptr != nullptr &&
-  //     "Failed to get address of clGetDeviceFunctionPointerINTEL function");
-
-  if (!func_ptr) {
-    if (function_pointer_ret)
+  // Get device supported extensions
+  size_t extension_size;
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS, 0, nullptr,
+                  &extension_size);
+  char *extensions = new char[extension_size];
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS,
+                  extension_size, extensions, nullptr);
+
+  // Check if clGetDeviceFunctionPointer is in list of extensions
+  if (is_in_separated_string(extensions, ' ',
+                             "clGetDeviceFunctionPointerINTEL")) {
+    delete[] extensions;
+
+    using FuncT = cl_int(CL_API_CALL *)(cl_device_id, cl_program, const char *,
+                                        cl_ulong *);
+
+    FuncT func_ptr =
+        reinterpret_cast<FuncT>(clGetExtensionFunctionAddressForPlatform(
+            cast<cl_platform_id>(platform), "clGetDeviceFunctionPointerINTEL"));
+
+    assert(func_ptr != nullptr &&
+           "Failed to get address of clGetDeviceFunctionPointerINTEL function");
+
+    pi_result piret_err = cast<pi_result>(
+        func_ptr(cast<cl_device_id>(device), cast<cl_program>(program),
+                 func_name, function_pointer_ret));
+    return piret_err;
+  } else {
+    // Use backup method to return placeholder address of 1 or 0 depending on if
+    // the function can be found this is needed by program_impl's has_kernel
+    delete[] extensions;
+    size_t Size;
+    cl_int Res =
+        clGetProgramInfo(cast<cl_program>(program),
+                         PI_PROGRAM_INFO_KERNEL_NAMES, 0, nullptr, &Size);
+    if (Res != CL_SUCCESS) {
       *function_pointer_ret = 0;
-    return PI_INVALID_DEVICE;
-  }
+      return cast<pi_result>(Res);
+    }
+
+    std::string ClResult(Size, ' ');
+    ret_err = clGetProgramInfo(cast<cl_program>(program),
+                               PI_PROGRAM_INFO_KERNEL_NAMES, ClResult.size(),
+                               &ClResult[0], nullptr);
+    if (Res != CL_SUCCESS) {
+      *function_pointer_ret = 0;
+      return cast<pi_result>(Res);
+    }
 
-  return cast<pi_result>(func_ptr(cast<cl_device_id>(device),
-                                  cast<cl_program>(program), func_name,
-                                  function_pointer_ret));
+    // Get rid of the null terminator and search for kernel_name
+    ClResult.pop_back();
+    *function_pointer_ret = is_in_separated_string(ClResult, ';', func_name);
+    return PI_SUCCESS;
+  }
```

Comment:
Clever solution! However, I fear that having the fallback mechanism at this level might be a problem for other parts of the runtime that uses `piextGetDeviceFunctionPointer` as it is returning invalid pointers. Granted it should fail if it ever gets to the fallback, it would be better to make the failure apparent after the call to `piextGetDeviceFunctionPointer` rather than whenever the returned pointer is used.

Therefore I think it would be better to have these functions return some identifiable `pi_result` (maybe `PI_EXTENSION_NOT_SUPPORTED` or something similar) and then fall back to this solution in `program_impl::has_kernel`. The CUDA and ROCm backend won't ever report `PI_EXTENSION_NOT_SUPPORTED` so they will never be affected by the fallback strategy.

Reply:
Ahh good idea! Ill make the changes to indicate that the extension is not working when the address is needed.

Reply:
> return some identifiable pi_result (maybe PI_EXTENSION_NOT_SUPPORTED

We could also add new result code to mean that the function is there (based on fallback search), but it's address cannot be obtained due to missing native support for such functionality

Reply:
I added two new pi_result's `PI_FALLBACK_SUCCESS` and `PI_FALLBACK_FAILURE` this is so the fallback method can indicate that it has failed and have set `getDeviceFunctionPointerImpl` to return an address of zero if a fallback method is used.